File: proselint/checks/dates_times/dates.py
"""
Dates.

---
layout:     post
source:     Garner's Modern American Usage
source_url: http://bit.ly/1T4alrY
title:      dates
date:       2014-06-10 12:31:19
categories: writing
---

Dates.

"""

import calendar

from proselint.registry.checks import Check, Padding, types

REGEX_MONTHS = Padding.SAFE_JOIN.format("|".join(calendar.month_name[1:]))

checks_decade_apostrophes = tuple(
    Check(
        check_type=types.ExistenceSimple(pattern=pattern),
        path="dates_times.dates.decade_apostrophes",
        message="Apostrophes aren't needed for decades.",
    )
    for pattern in (r"\d0\'s", r"\d\d\d0\'s")
)

check_dash_and_from = Check(
    check_type=types.ExistenceSimple(pattern=r"from \d+[^ \t\n\r\f\v\w_\.]\d+"),
    path="dates_times.dates.dash_and_from",
    message="When specifying a date range, write 'from X to Y'.",
)

check_month_year_comma = Check(
    check_type=types.ExistenceSimple(pattern=REGEX_MONTHS + r", \d{3,}"),
    path="dates_times.dates.month_year_comma",
    message="When specifying a month and year, no comma is needed.",
)

check_month_of_year = Check(
    check_type=types.ExistenceSimple(pattern=REGEX_MONTHS + r" of \d{3,}"),
    path="dates_times.dates.month_of_year",
    message="When specifying a month and year, 'of' is unnecessary.",
)

__register__ = (
    *checks_decade_apostrophes,
    check_dash_and_from,
    check_month_year_comma,
    check_month_of_year,
)

File: proselint-rs/crates/proselint-checks/src/dates_times/dates.rs
use proselint_registry::{pad, checks::{Check, CheckType, Padding}};
use const_format::concatcp;

pub const EXAMPLES_PASS: &[&str] = &[
	"Smoke phrase with nothing flagged.",
	"It happened in the 90s.",
	"It happened in the 1980s.",
	"It happened from 2000 to 2005.",
	"It happened in August 2008.",
	"It happened in August 2008.",
	"Dr. Dre suggested to 50's manager that he look into signing \
	Eminem to the G-Unit record label.",
];
pub const EXAMPLES_FAIL: &[&str] = &[
	"It happened in the 90's.",
	"It happened in the 1980's.",
	"It happened from 2000-2005.",
	"It happened in August, 2008.",
	"It happened in August of 2008.",
	"The 50's were swell.",
	"From 1999-2002, Sally served as chair of the committee.",
];

const CHECK_DECADES_APOSTROPHES_SHORT: Check = Check {
	check_type: CheckType::ExistenceSimple {
		pattern: pad!(Padding::WordsInText, r"\d0\'s"),
		exceptions: &[],
	},
	path: "dates_times.dates.apostrophes",
	msg: "Apostrophes aren't needed for decades.",
	..Check::default()
};

const CHECK_DECADES_APOSTROPHES_LONG: Check = Check {
	check_type: CheckType::ExistenceSimple {
		pattern: pad!(Padding::WordsInText, r"\d\d\d0\'s"),
		exceptions: &[],
	},
	path: "dates_times.dates.apostrophes",
	msg: "Apostrophes aren't needed for decades.",
	..Check::default()
};

const CHECK_DASH_AND_FROM: Check = Check {
	check_type: CheckType::ExistenceSimple {
		pattern: pad!(Padding::WordsInText, r"from \d+[^ \t\n\r\f\v\w.]\d+"),
		exceptions: &[],
	},
	path: "dates_times.dates.dash_and_from",
	msg: "When specifying a date range, write 'from X to Y'.",
	..Check::default()
};

// the Rust standard library does not have a list of months, nor a way to join
// instead of concat slices in a const context.
const MONTHS_SEPARATED: &str =
	"January|February|March|April|May|June|July \
	|August|September|October|November|December";

const CHECK_MONTH_YEAR_COMMA: Check = Check {
	check_type: CheckType::ExistenceSimple {
		pattern: concatcp!(pad!(Padding::SafeJoin, MONTHS_SEPARATED), r", \d{3,}"),
		exceptions: &[],
	},
	path: "dates_times.dates.month_year_comma",
	msg: "When specifying a month and year, no comma is needed.",
	..Check::default()
};

const CHECK_MONTH_OF_YEAR: Check = Check {
	check_type: CheckType::ExistenceSimple {
		pattern: concatcp!(pad!(Padding::SafeJoin, MONTHS_SEPARATED), r"of \d{3,}"),
		exceptions: &[],
	},
	path: "dates_times.dates.month_of_year",
	msg: "When specifying a month and year, 'of' is unnecessary.",
	..Check::default()
};

pub const REGISTER: &[Check] = &[
	CHECK_DECADES_APOSTROPHES_SHORT,
	CHECK_DECADES_APOSTROPHES_LONG,
	CHECK_DASH_AND_FROM,
	CHECK_MONTH_YEAR_COMMA,
	CHECK_MONTH_OF_YEAR,
];
